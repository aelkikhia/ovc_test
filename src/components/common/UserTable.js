import React from 'react'
import {connect} from 'react-redux';
import {fetchUsers} from "../../actions/users";

const TABLE_HEADERS = ['Name', 'Email', 'City', 'Company'];

export class UserTable extends React.Component {

    componentDidMount() {
        this.props.fetchUsers();
    }

    renderTableHeader () {
        return TABLE_HEADERS.map((name, index) => {
            return (<th key={index}>{name}</th>);
        })
    }

    renderTableData() {

        return this.props.users.users.map((user, index) =>  {
            const { name, email, address, company } = user;
            return (
                <tr key={index}>
                    <td data-label='name'>{name}</td>
                    <td data-label='email'><a href={`mailto:${email}`}>{email}</a></td>
                    <td data-label='city'>{address.city}</td>
                    <td data-label='company'>{company.name}</td>
                </tr>
            );
        });
    }

    renderTable() {
        return (
            <table className="ui celled table">
                <thead><tr>{this.renderTableHeader()}</tr></thead>
                <tbody>{this.renderTableData()}</tbody>
            </table>
        );
    }

    render() {
        if(this.props.users.error) {
            return <div>A network error occured</div>
        }
        if(this.props.users.loading) {
            return <div>Loading...</div>
        }
        return this.renderTable();
    }
}

const mapStateToProps = state => ({
    ...state
});

const mapDispatchToProps = dispatch => ({
    fetchUsers: () => dispatch(fetchUsers())
});

export default connect(mapStateToProps, mapDispatchToProps)(UserTable)