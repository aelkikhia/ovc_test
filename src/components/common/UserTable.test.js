import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import {UserTable} from './UserTable';

configure({ adapter: new Adapter() });

describe('UserTable Component', () => {
    let props;

    beforeEach(() => {
        props = {
            fetchUsers: jest.fn(),
            users: {
                users: [],
                loading: false,
                error: false
            }
        };
    });

    it('renders without crashing', () => {
        const tree = renderer.create(<UserTable {...props} />);

        expect(tree.toJSON()).toMatchSnapshot();
    });

    it('renders an error message when a network error occurs', () => {
        props.users.error = true;
        const tree = renderer.create(<UserTable {...props} />);

        expect(tree.toJSON()).toMatchSnapshot();
    });

    it('renders an error message when a network error occurs', () => {
        props.users.loading = true;
        const tree = renderer.create(<UserTable {...props} />);

        expect(tree.toJSON()).toMatchSnapshot();
    });

    it('renders the User component correctly', () => {
        props.users.users = [
            {
                "name": "Leanne Graham",
                "email": "Sincere@april.biz",
                "address": {"city": "Gwenborough",},
                "company": {"name": "Romaguera-Crona",}
            }
        ];
        const wrapper = shallow(<UserTable {...props} />);

        expect(wrapper.find('td').length).toBe(4);
    });
});
