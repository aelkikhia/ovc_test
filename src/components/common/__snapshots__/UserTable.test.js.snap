// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserTable Component renders an error message when a network error occurs 1`] = `
<div>
  A network error occured
</div>
`;

exports[`UserTable Component renders an error message when a network error occurs 2`] = `
<div>
  Loading...
</div>
`;

exports[`UserTable Component renders without crashing 1`] = `
<table
  className="ui celled table"
>
  <thead>
    <tr>
      <th>
        Name
      </th>
      <th>
        Email
      </th>
      <th>
        City
      </th>
      <th>
        Company
      </th>
    </tr>
  </thead>
  <tbody />
</table>
`;
